### ISSUE TITLE
Use PyO3

### ISSUE THREAD

**Issue by danielabrozzoni** [bitcoindevkit/rust-hwi#8]
Instead of just launching the hwi binary with python, call directly its functions using PyO3

**Comment by wszdexdrf** [comment]
I think the easiest way would be to import `hwilib` using PyO3 and then call the functions as needed. But PyO3 import function returns a `&PyModule` with the  lifetime as long as the Python GIL is held. So here we have a choice between 2 _very_ bad options. Either import the module needed every single function call, or hold the GIL the entire time. I just wanted your opinion on this.
FYI, importing `hwilib.commands` took around 0.01s on my machine. Not rigorously tested, though.

**Comment by afilini** [comment]
I think we could get away with holding the GIL the entire time, since I don't think hwi spawns threads that need to run independently from our calls.

I would try with that approach first and if it fails we can try something else.

**Comment by wszdexdrf** [comment]
That problem was solved by converting `&PyModule` to `Py<PyModule>`. But now here is a puzzling problem:
```
fn test_custom() {
        let lib = HWILib::initialize().unwrap();
        let device = get_first_device(&lib);
        let account = Some(10);
        device.get_descriptors(account, &lib).unwrap();
        HWIDevice::enumerate(&lib).unwrap();
}
```
This test fails. If I comment out any one of the lines  `device.get_descriptors(account, &lib).unwrap();` or `HWIDevice::enumerate(&lib).unwrap();`, the test passes. Also, all the preexisting tests pass if I run them one at a time but fail if ran all at once with `cargo test`. I receive a `Error` as follows:
```
HWIError("Error Error(\"missing field `needs_pin_sent`\", line: 0, column: 0) while deserializing Array([Object({\"code\": Number(-13), \"error\": String(\"Could not open client or get fingerprint information: open failed\"), \"label\": Null, \"model\": String(\"ledger_nano_s\"), \"path\": String(\"0001:0009:00\"), \"type\": String(\"ledger\")})])")'
```
Is that some problem with my hwwallet's implementation of HWI?

**Comment by wszdexdrf** [comment]
Apparently the error thrown is [this one](https://github.com/bitcoin-core/HWI/blob/master/hwilib/devices/ledger.py#L137)

**Comment by wszdexdrf** [comment]
That problem was solved by converting `&PyModule` to `Py<PyModule>`, where the latter is GIL independent. Because of that, we can hold the GIL only during the function calls, and still don't need to import libraries. Thanks for the help, though. 
There is a small problem though, now. All tests pass when executed one at a time but fail to pass when run using `cargo test` I am guessing that has to do with rust running tests in parallel. Can we ignore that problem exists? 😆

**Comment by afilini** [comment]
Ah that's great news then!

If running multiple tests at once makes them fail, maybe we should add mutexes somewhere to prevent multiple threads in the same binary to make weird things happen..

For the tests in particular it's not that important, we could just run them sequentially. But I'm afraid the same issues could happen in running binaries if we get multiple hwi calls concurrently.

**Comment by anishagg17** [comment]
@danielabrozzoni if we import hwilib(which serves APIs in python via commands module) we would need to have a copy of `hwilib` folder  as hwlib isn't a python package available. This would make code implementation messy as we would have to manually sync hwilib folder from the source HWI. Please correct me if I am wrong, suggestions are most welcome 🤗

**Comment by danielabrozzoni** [comment]
> as hwlib isn't a python package available

It's called `hwi`: https://pypi.org/project/hwi/

**Comment by wszdexdrf** [comment]
> This would make code implementation messy. 
Not really. It is actually a good thing since an update in the source won't break this library. Also, minimal impact on the user's system. Actual code implementation is very similar irrespective of using pip installed library or using the source (see [from_code()](https://docs.rs/pyo3/latest/pyo3/prelude/struct.PyModule.html#method.from_code)).

**Comment by danielabrozzoni** [comment]
> I think it's just a cli tool and doesn't expose the actual python functions?

What happens if you `pip install hwi` and then import `hwilib`?

**Comment by danielabrozzoni** [comment]
Closed by #19
